component/newstorage/batchedstore/batchedstore.go

```diff
@@ -0,0 +1,348 @@
+		Key:   key,
+		Value: value,
+		Tags:  tags,
+	})
+
+	currentBatchLength := len(s.currentBatch)
+
+	s.Unlock()
+
+	if currentBatchLength >= s.batchSizeLimit {
+		err := s.Flush()
+		if err != nil {
+			return fmt.Errorf(failFlush, err)
+		}
+	}
+
+	return nil
+}
+
+// TODO (#2485): Check data in current batch before resorting to a flush.
+func (s *store) Get(key string) ([]byte, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	value, err := s.underlyingStore.Get(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get value from underlying store: %w", err)
+	}
+
+	return value, nil
+}
+
+func (s *store) GetTags(key string) ([]newstorage.Tag, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	tags, err := s.underlyingStore.GetTags(key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tags from underlying store: %w", err)
+	}
+
+	return tags, nil
+}
+
+func (s *store) GetBulk(keys ...string) ([][]byte, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	values, err := s.underlyingStore.GetBulk(keys...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get values from underlying store: %w", err)
+	}
+
+	return values, nil
+}
+
+func (s *store) Query(expression string, options ...newstorage.QueryOption) (newstorage.Iterator, error) {
+	err := s.Flush()
+	if err != nil {
+		return nil, fmt.Errorf(failFlush, err)
+	}
+
+	iterator, err := s.underlyingStore.Query(expression, options...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get values from underlying store: %w", err)
+	}
+
+	return iterator, nil
+}
+
+func (s *store) Delete(key string) error {
+	if key == "" {
+		return errEmptyKey
+	}
+
+	s.Lock()
+
+	s.currentBatch = append(s.currentBatch, newstorage.Operation{
+		Key:   key,
+		Value: nil,
+	})
+
+	currentBatchLength := len(s.currentBatch)
+
+	s.Unlock()
+
+	if currentBatchLength >= s.batchSizeLimit {
+		err := s.Flush()
+		if err != nil {
+			return fmt.Errorf(failFlush, err)
+		}
+	}
+
+	return nil
+}
+
+func (s *store) Batch(operations []newstorage.Operation) error {
+	for _, operation := range operations {
+		if operation.Key == "" {
+			return errors.New("an operation's key was empty")
+		}
+	}
+
+	for _, operation := range operations {
+		s.Lock()
+
+		s.currentBatch = append(s.currentBatch, operation)
+
+		currentBatchLength := len(s.currentBatch)
+
+		s.Unlock()
+
+		if currentBatchLength >= s.batchSizeLimit {
+			err := s.Flush()
+			if err != nil {
+				return fmt.Errorf(failFlush, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *store) Close() error {
+	s.close(s.name)
+
+	err := s.Flush()
+	if err != nil {
+		return fmt.Errorf(failFlush, err)
+	}
+
+	err = s.underlyingStore.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close underlying store: %w", err)
+	}
+
+	return nil
+}
+
+func (s *store) Flush() error {
```

Comment:
@DRK3 @bstasyszyn multiple concurrent `Put()`, `Delete()`, and `Batch()` calls can cause `Flush()` to be executed multiple times in parallel, and if another `Put()` or `Batch()` partially fills `s.currentBatch` after one of those Flushes empties it, then the next `Flush()` to enter its critical section would flush `s.currentBatch` before it's been filled again, when the intention behind the call to `Flush()` was to flush a full buffer. 

It's not a correctness issue, all concurrent calls to `Put()`, `Delete()` and `Batch()` will have each of their operations flushed exactly once, so I'm not sure it's a concern, but it will occasionally spread operations out across extraneous flushes. Might be a performance concern depending on how we use it.

Reply:
@Moopli Thanks for catching this. Also as you pointed out in a separate chat, it would be possible for log statements to be inappropriately printed out when there are no operations to flush.

I just pushed a fix to this PR that should address that. Let me know how it looks @Moopli @bstasyszyn